src/psm/src/ir_network.cpp

```diff
@@ -0,0 +1,1299 @@
+
+        nodes_[layer].push_back(std::move(center));
+      }
+    }
+
+    if (has_routing_term) {
+      iterm_nodes_.push_back(std::move(base_node));
+    }
+  }
+
+  return shapes_by_layer;
+}
+
+IRNetwork::LayerMap<IRNetwork::Polygon90Set>
+IRNetwork::generatePolygonsFromBTerms(std::vector<TerminalNode*>& terminals)
+{
+  using boost::polygon::operators::operator+=;
+
+  const utl::DebugScopedTimer timer(
+      logger_, utl::PSM, "timer", 1, "Generate shapes from BTerms: {}");
+
+  LayerMap<Polygon90Set> shapes_by_layer;
+
+  for (auto* bterm : net_->getBTerms()) {
+    for (auto* bpin : bterm->getBPins()) {
+      for (auto* geom : bpin->getBoxes()) {
+        for (const auto& [layer, shapes] :
+             generatePolygonsFromBox(geom, odb::dbTransform())) {
+          shapes_by_layer[layer] += shapes;
+        }
+
+        if (geom->isVia()) {
+          continue;
+        }
+
+        auto* layer = geom->getTechLayer();
+        const odb::Rect pin_shape = geom->getBox();
+        const odb::Point center(pin_shape.xCenter(), pin_shape.yCenter());
+
+        // create bpin nodes
+        auto term = std::make_unique<TerminalNode>(pin_shape, layer);
+        auto pin_node = std::make_unique<BPinNode>(bpin, pin_shape, layer);
+
+        connections_.push_back(
+            std::make_unique<TermConnection>(term.get(), pin_node.get()));
+
+        nodes_[layer].push_back(std::move(term));
+        bpin_nodes_.push_back(std::move(pin_node));
+      }
+    }
+  }
+
+  return shapes_by_layer;
+}
+
+void IRNetwork::processPolygonToRectangles(
+    odb::dbTechLayer* layer,
+    const IRNetwork::Polygon90& polygon,
+    const IRNetwork::TerminalTree& terminals,
+    std::vector<std::unique_ptr<Shape>>& new_shapes,
+    std::vector<std::unique_ptr<Node>>& new_nodes,
+    std::map<Shape*, std::set<Node*>>& terminal_connections)
+{
+  using boost::polygon::operators::operator+=;
+  using Rectangle = boost::polygon::rectangle_data<int>;
+
+  auto get_layer_orientation
+      = [](odb::dbTechLayer* layer) -> boost::polygon::orientation_2d {
+    switch (layer->getDirection()) {
+      case odb::dbTechLayerDir::NONE:
+      case odb::dbTechLayerDir::HORIZONTAL:
+        return boost::polygon::orientation_2d(
+            boost::polygon::orientation_2d_enum::VERTICAL);
+        break;
+      case odb::dbTechLayerDir::VERTICAL:
+        return boost::polygon::orientation_2d(
+            boost::polygon::orientation_2d_enum::HORIZONTAL);
+    }
+    return boost::polygon::orientation_2d(
+        boost::polygon::orientation_2d_enum::VERTICAL);
+  };
+
+  Polygon90Set shape_poly_set;
+  shape_poly_set += polygon;
+
+  std::vector<Rectangle> rect_shapes;
+  shape_poly_set.get_rectangles(rect_shapes, get_layer_orientation(layer));
+
+  std::set<odb::Rect> search_rect_shapes;
+  for (const auto& poly_rect : rect_shapes) {
+    search_rect_shapes.emplace(
+        xl(poly_rect), yl(poly_rect), xh(poly_rect), yh(poly_rect));
+  }
```

Comment:
With `#include "odb/geom_boost.h"` this can be reduced to 

````
  std::vector<odb::Rect> search_rect_shapes;
  shape_poly_set.get_rectangles(search_rect_shapes, get_layer_orientation(layer));
````

[auto for the iterator type below avoids needing to update them]

Reply:
thats awesome (this will appear in a lot of places)

Reply:
When I tried I get:
```
/usr/local/include/boost/polygon/polygon_90_set_data.hpp: In instantiation of ‘void boost::polygon::polygon_90_set_data<T>::get_rectangles(output_container&, boost::polygon::orientation_2d) const [with output_container = std::set<odb::Rect>; T = int]’:
..../OpenROAD/src/psm/src/ir_network.cpp:371:32:   required from here
/usr/local/include/boost/polygon/polygon_90_set_data.hpp:193:38: error: no matching function for call to ‘transform(const odb::Rect&, boost::polygon::axis_transformation)’
  193 |           ::boost::polygon::transform(*itr, axis_transformation(axis_transformation::SWAP_XY));`
```

Reply:
I just made the change and pushed it to your branch.